### ISSUE TITLE
Add ability to download all certificates in the report as a teacher and bulk download all site certificates as an administrator

### ISSUE THREAD

**Issue by Markibacsi** [mdjnelson/moodle-mod_customcert#205]
Hi Mark

Great work on the plugin - a fantastic resource.

Our teachers are focused on the delivery of their classes and the admin side is dealt with by our admin team. Part of our service is to provide a printed certificate to our students at the end of the course, and we usually have about 20-30 courses finishing at around the same time.

Currently we split a list of the courses, enter each course and download the certificates one by one. Would it be possible to add a 'download all' button to grab all issued certificates site-wide between two dates?

Thanks

**Comment by Simona77** [comment]
Hi mark, I would like to know if you succeeeded in adding the "download all" button. I've got the same issue described by Markibacsi. Please let me know.

**Comment by Simona77** [comment]
Hi Mark, I would like to know if you succeeeded in adding the "download all" button. I've got the same issue described by Markibacsi. Please let me know.

**Comment by pavlorenc** [comment]
Hi Mark, so Bulk certificates export it's no longer a planned feature? :-/

**Comment by markn86** [comment]
Sorry Pavel, I just didn't have the time to do it. I have left Moodle HQ and am currently doing contract work which is keeping me busy. However, after nearly 10 years of Moodle development I am keen to move on to something else, so am currently in the process of finding work elsewhere. However, I would be more than happy to review a patch and integrate it. :)

**Comment by aguinagadan** [comment]
Hi, can anyone help me to implement this function? I am willing to pay for your time and also share it with the community. @mdjnelson

**Comment by mdjnelson** [comment]
I personally do not have as much time as I want to work on this as I no longer work for Moodle HQ. I have had good experiences with the Moodle partner Catalyst IT who have developed features for this plugin before. I would suggest getting in contact with them. Thanks.

**Comment by ricardoh34** [comment]
Hello, 

I´m interested also in this feature, maybe we could work in it, @mdjnelson could you guide me a bit in the structure of the module? Where can I start to review this issue? It is really important for us. Is there anyone else working on this module?

**Comment by mdjnelson** [comment]
There is currently no one else except myself working on this module and I do it in my free time. If you want you could contact https://catalyst-au.net/contact-us to get a quote on the work required, otherwise you will have to wait for me to implement the feature. Sorry.

**Comment by aguinagadan** [comment]
Hello everyone, with my team we managed to implement that option.
Save a copy on the service when the certificate is generated. I have no problem sharing with the community if mdjnelson allows it.

**Comment by ricardoh34** [comment]
I would greatly appreciate your improvement!

Thanks!

**Comment by mdjnelson** [comment]
@aguinagadan of course you are welcome to share. :)

**Comment by nautiluspro** [comment]
I'm also interested in this topic. I'm keen on it...

**Comment by harismateen** [comment]
> Hello everyone, with my team we managed to implement that option.
> Save a copy on the service when the certificate is generated. I have no problem sharing with the community if mdjnelson allows it.

Can share if it's avaliable?

**Comment by mdjnelson** [comment]
Since there is such a high demand for this I suggest seeing if you can sort out some funding to implement this. As listed above I would contact Catalyst (Moodle partner) at https://catalyst-au.net/contact-us.

**Comment by rgroba** [comment]
Until they implement it.
You can try downloading with jquery in the chrome console


```
//obtener todos los enlaces
var divs = $("a > i.fa-download");
var counter = 0;
//repetimos el proceso por todos los nodos hijos 
setInterval(function() {
	
	if(counter >= divs.length){
			return '';
	}
	//otenemos el hijo en el que estamos en ese momento
	var div = $(divs[counter]);
	//añadir función para que descarge directamente y no por cree un enlace nuevo
    div.parent().attr('download','download');
	//cuando clicamos cambiamos el icono
	div.on('click', function() {
		$(this).removeClass("fa-download");
		$(this).addClass("fa-solid");
		$(this).addClass("fa-check");	
	});
		
	//pulsamos el botón para descargar
	div.trigger('click');

	counter ++;
	

},3000);

```

**Comment by adanisyrex** [comment]
> Until they implement it. You can try downloading with jquery in the chrome console
> 
> ```
> //obtener todos los enlaces
> var divs = $("a > i.fa-download");
> var counter = 0;
> //repetimos el proceso por todos los nodos hijos 
> setInterval(function() {
> 	
> 	if(counter >= divs.length){
> 			return '';
> 	}
> 	//otenemos el hijo en el que estamos en ese momento
> 	var div = $(divs[counter]);
> 	//añadir función para que descarge directamente y no por cree un enlace nuevo
>     div.parent().attr('download','download');
> 	//cuando clicamos cambiamos el icono
> 	div.on('click', function() {
> 		$(this).removeClass("fa-download");
> 		$(this).addClass("fa-solid");
> 		$(this).addClass("fa-check");	
> 	});
> 		
> 	//pulsamos el botón para descargar
> 	div.trigger('click');
> 
> 	counter ++;
> 	
> 
> },3000);
> ```

Wonderful! Gracias!

**Comment by adanisyrex** [comment]
rgroba: Wonderfu!! Gracias!

**Comment by aguinagadan** [comment]
Hello, sorry for the disconnection.

(sorry for my english)

We solved it in the following way:

We store the file locally in the "/mod/customcert/files/" folder taking advantage of the user's action when downloading their certificate.

To achieve this, we add the code in line 325 of the file "/mod/customcert/classes/template.php" (see image with code)


![image](https://user-images.githubusercontent.com/7682069/175183046-9c657b11-643b-406d-ab00-02b18d17340c.png)


Then, a file must be created to execute the download including the course id in the url, it would be something similar to this: "https://www.mymoodle.com/local/downloadCerts.php?idCurso=40"

(Be careful with the route on line 6)

![image](https://user-images.githubusercontent.com/7682069/175184353-35324b8e-fcd1-402d-bf67-3c6a45a8951a.png)
![image](https://user-images.githubusercontent.com/7682069/175184391-b6026335-8e97-40f0-95c6-af4918cd7453.png)
![image](https://user-images.githubusercontent.com/7682069/175184518-67983f71-87f8-47a2-9ecb-84c6609180f2.png)
![image](https://user-images.githubusercontent.com/7682069/175184585-3e244aac-4739-41e6-96b8-c337ccbcbb73.png)


I hope it's not too late and it helps.

Greetings to the community.

by aguinagadan

**Comment by JaderGabriel** [comment]
> Hello, sorry for the disconnection.
> 
> (sorry for my english)
> 
> We solved it in the following way:
> 
> We store the file locally in the "/mod/customcert/files/" folder taking advantage of the user's action when downloading their certificate.
> 
> To achieve this, we add the code in line 325 of the file "/mod/customcert/classes/template.php" (see image with code)
> 
> ![image](https://user-images.githubusercontent.com/7682069/175183046-9c657b11-643b-406d-ab00-02b18d17340c.png)
> 
> Then, a file must be created to execute the download including the course id in the url, it would be something similar to this: "https://www.mymoodle.com/local/downloadCerts.php?idCurso=40"
> 
> (Be careful with the route on line 6)
> 
> ![image](https://user-images.githubusercontent.com/7682069/175184353-35324b8e-fcd1-402d-bf67-3c6a45a8951a.png) ![image](https://user-images.githubusercontent.com/7682069/175184391-b6026335-8e97-40f0-95c6-af4918cd7453.png) ![image](https://user-images.githubusercontent.com/7682069/175184518-67983f71-87f8-47a2-9ecb-84c6609180f2.png) ![image](https://user-images.githubusercontent.com/7682069/175184585-3e244aac-4739-41e6-96b8-c337ccbcbb73.png)
> 
> I hope it's not too late and it helps.
> 
> Greetings to the community.
> 
> by aguinagadan

Morning,
Do this working ? Could you share de print or imagem about result in moodle? Thanks

**Comment by adanisyrex** [comment]
> > Hello, sorry for the disconnection.
> > (sorry for my english)
> > We solved it in the following way:
> > We store the file locally in the "/mod/customcert/files/" folder taking advantage of the user's action when downloading their certificate.
> > To achieve this, we add the code in line 325 of the file "/mod/customcert/classes/template.php" (see image with code)
> > ![image](https://user-images.githubusercontent.com/7682069/175183046-9c657b11-643b-406d-ab00-02b18d17340c.png)
> > Then, a file must be created to execute the download including the course id in the url, it would be something similar to this: "https://www.mymoodle.com/local/downloadCerts.php?idCurso=40"
> > (Be careful with the route on line 6)
> > ![image](https://user-images.githubusercontent.com/7682069/175184353-35324b8e-fcd1-402d-bf67-3c6a45a8951a.png) ![image](https://user-images.githubusercontent.com/7682069/175184391-b6026335-8e97-40f0-95c6-af4918cd7453.png) ![image](https://user-images.githubusercontent.com/7682069/175184518-67983f71-87f8-47a2-9ecb-84c6609180f2.png) ![image](https://user-images.githubusercontent.com/7682069/175184585-3e244aac-4739-41e6-96b8-c337ccbcbb73.png)
> > I hope it's not too late and it helps.
> > Greetings to the community.
> > by aguinagadan
> 
> Morning, Do this working ? Could you share de print or imagem about result in moodle? Thanks

I didn´t try it, I found super easy the way of rgroba, check his post of 6th June, it works wonderfull

**Comment by GonDragon** [comment]
Hi.

In the place I'm working now, they required a function to bulk download all the certificates from a course. I made a small local plugin to add a button "Download Certificates" to the settings menu of every course that contains an user with at least 1 emmited certificate. The certificates are all together inside a ZIP file.

The plugin is not extensively tested, but maybe it could be useful for somebody. This is link to the repo:
[bulkcustomcert](https://github.com/GonDragon/bulkcustomcert)

**Comment by adanisyrex** [comment]
Hi Gonzalo,

 

Wonderful, thanks for going into that. I have installed in a test version I have for trying new stuffs before installing in my production sites and I didn’t get it to work I think, I don´t see the button you refer in the setting menu. Anyway I am understanding that your local plugin, downloads all the certificates that a user itself has gotten? What I was needing is to download in an action all the certificates I filter in the list/ table (quiz/report.php) without having to go one by one. I got that with Google Chrome and the code one user was giving. Is it that what you are trying to get with the plugin?

 

I am using Moodle 3.5  Custom Certificate 3.5.8 (2018051709)

 

Thanks for your attention, warm regards, Adán

 

De: Gonzalo Romero ***@***.***> 
Enviado el: martes, 29 de noviembre de 2022 23:38
Para: mdjnelson/moodle-mod_customcert ***@***.***>
CC: adanisyrex ***@***.***>; Comment ***@***.***>
Asunto: Re: [mdjnelson/moodle-mod_customcert] Download all certificates as an administrator (#205)

 

Hi.

In the place I'm working now, they required a function to bulk download all the certificates from a course. I made a small local plugin to add a button "Download Certificates" to the settings menu of every course that contains an user with at least 1 emmited certificate. The certificates are all together inside a ZIP file.

The plugin is not extensively tested, but maybe it could be useful for somebody. This is link to the repo:
bulkcustomcert <https://github.com/GonDragon/bulkcustomcert>

**Comment by GonDragon** [comment]
No, the plugin does not download all the certificates from one user. What it does, is that it scans one Course, and download all the certificates from the users of that course. The reason of why I made this plugin, is because my client required a quick way to download all the certificates emited in that course, so they can send emails to the students without having to go profile by profile, downloading the certificate.

To download all the certs from one specific user, I think the best option is still the script you mentioned.

Here a picture of how the "Download all the certificates" should appear on a course:
![Screenshot_1](https://user-images.githubusercontent.com/7538144/204928170-df67ed5e-7146-452b-a709-17b25458e991.png)

The link points to "http://your-mooodle.com/local/bulkcustomcert/index.php?id={courseid}" and you should have the "mod/customcert:viewallcertificates" capability to download them.

**Comment by Jmoramun** [comment]
> 

Hi Gonzalo, 
I tested the module, as an administrator with manager  and teacher role inside a few courses, with the viewallcertificates capability, and the button it's not even appearing. 

Is there any requisites about customcert or moodle version?

Thanks in advance!


```

**Comment by ShubhamNCRB** [comment]
> > Hello, sorry for the disconnection.
> > (sorry for my english)
> > We solved it in the following way:
> > We store the file locally in the "/mod/customcert/files/" folder taking advantage of the user's action when downloading their certificate.
> > To achieve this, we add the code in line 325 of the file "/mod/customcert/classes/template.php" (see image with code)
> > ![image](https://user-images.githubusercontent.com/7682069/175183046-9c657b11-643b-406d-ab00-02b18d17340c.png)
> > Then, a file must be created to execute the download including the course id in the url, it would be something similar to this: "https://www.mymoodle.com/local/downloadCerts.php?idCurso=40"
> > (Be careful with the route on line 6)
> > ![image](https://user-images.githubusercontent.com/7682069/175184353-35324b8e-fcd1-402d-bf67-3c6a45a8951a.png) ![image](https://user-images.githubusercontent.com/7682069/175184391-b6026335-8e97-40f0-95c6-af4918cd7453.png) ![image](https://user-images.githubusercontent.com/7682069/175184518-67983f71-87f8-47a2-9ecb-84c6609180f2.png) ![image](https://user-images.githubusercontent.com/7682069/175184585-3e244aac-4739-41e6-96b8-c337ccbcbb73.png)
> > I hope it's not too late and it helps.
> > Greetings to the community.
> > by aguinagadan
> 
> Morning, Do this working ? Could you share de print or imagem about result in moodle? Thanks

How do i execute the query(Jquery)? Please help.

**Comment by shikizisykes26** [comment]
> No, the plugin does not download all the certificates from one user. What it does, is that it scans one Course, and download all the certificates from the users of that course. The reason of why I made this plugin, is because my client required a quick way to download all the certificates emited in that course, so they can send emails to the students without having to go profile by profile, downloading the certificate.
> 
> To download all the certs from one specific user, I think the best option is still the script you mentioned.
> 
> Here a picture of how the "Download all the certificates" should appear on a course: ![Screenshot_1](https://user-images.githubusercontent.com/7538144/204928170-df67ed5e-7146-452b-a709-17b25458e991.png)
> 
> The link points to "[http://your-mooodle.com/local/bulkcustomcert/index.php?id={courseid}](http://your-mooodle.com/local/bulkcustomcert/index.php?id=%7Bcourseid%7D)" and you should have the "mod/customcert:viewallcertificates" capability to download them.

hello @GonDragon 
i use moodle 4.1 then had install your plugin, but when i check can't find the button on course page even use site administrator role.
![image](https://github.com/mdjnelson/moodle-mod_customcert/assets/33312950/a1b3860c-2f69-4847-b852-510c60382f79)


I open the link you provided before (already changed the moodle site and course id) nothing happens and it just returns to the course page. kindly please advise

**Comment by ngonguyentuanhhon** [comment]
> > Không, plugin không tải xuống tất cả chứng chỉ từ một người dùng. Những gì nó làm là quét một Khóa học và tải xuống tất cả các chứng chỉ từ người dùng của khóa học đó. Lý do tại sao tôi tạo plugin này là vì khách hàng của tôi yêu cầu một cách nhanh chóng để tải xuống tất cả các chứng chỉ được cấp trong khóa học đó, để họ có thể gửi email cho sinh viên mà không cần phải xem từng hồ sơ, tải xuống chứng chỉ.
> > Để tải xuống tất cả các chứng chỉ từ một người dùng cụ thể, tôi nghĩ tùy chọn tốt nhất vẫn là tập lệnh bạn đã đề cập.
> > Dưới đây là hình ảnh về cách "Tải xuống tất cả các chứng chỉ" sẽ xuất hiện trên một khóa học:![Ảnh chụp màn hình_1](https://user-images.githubusercontent.com/7538144/204928170-df67ed5e-7146-452b-a709-17b25458e991.png)
> > Liên kết trỏ đến " [http://your-mooodle.com/local/bulkcustomcert/index.php?id={courseid}](http://your-mooodle.com/local/bulkcustomcert/index.php?id=%7Bcourseid%7D) " và bạn phải có khả năng "mod/customcert:viewallcertificates" để tải chúng xuống.
> 
> Xin chào@GonDragon tôi sử dụng moodle 4.1 sau đó đã cài đặt plugin của bạn, nhưng khi tôi kiểm tra thì không thể tìm thấy nút trên trang khóa học ngay cả khi sử dụng vai trò quản trị viên trang web. ![hình ảnh](https://private-user-images.githubusercontent.com/33312950/265868009-a1b3860c-2f69-4847-b852-510c60382f79.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0NTQ3NjQsIm5iZiI6MTcwNTQ1NDQ2NCwicGF0aCI6Ii8zMzMxMjk1MC8yNjU4NjgwMDktYTFiMzg2MGMtMmY2OS00ODQ3LWI4NTItNTEwYzYwMzgyZjc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE3VDAxMjEwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwODdkYjhiYjEwNWU2MGUyMDViY2RmM2VlNmZlZTVmMWNjYzQ3Y2M3NWZjZGNjZTBjYzMwOTFhNjcyYzhhNDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6iyv5uo1RoA7QcMPjupjZ6p1g8D8DfTolnj5aTURfB0)
> 
> Tôi mở liên kết bạn cung cấp trước đó (đã thay đổi trang moodle và id khóa học) không có gì xảy ra và nó chỉ quay lại trang khóa học. xin vui lòng tư vấn


I corrected a few places in @GonDragon code and the plugin worked very well. The problem is that there is confusion between role_assignments.id and users.id in code. I think it's probably due to the difference in Moodle version

If @GonDragon  agrees I can send you the fix.

**Comment by mdjnelson** [comment]
Can someone create a PR?

**Comment by gustavoheredia21** [comment]
> > > Không, plugin không tải xuống tất cả chứng chỉ từ một người dùng. Những gì nó làm là quét một Khóa học và tải xuống tất cả các chứng chỉ từ người dùng của khóa học đó. Lý do tại sao tôi tạo plugin này là vì khách hàng của tôi yêu cầu một cách nhanh chóng để tải xuống tất cả các chứng chỉ được cấp trong khóa học đó, để họ có thể gửi email cho sinh viên mà không cần phải xem từng hồ sơ, tải xuống chứng chỉ.
> > > Để tải xuống tất cả các chứng chỉ từ một người dùng cụ thể, tôi nghĩ tùy chọn tốt nhất vẫn là tập lệnh bạn đã đề cập.
> > > Dưới đây là hình ảnh về cách "Tải xuống tất cả các chứng chỉ" sẽ xuất hiện trên một khóa học:![Ảnh chụp màn hình_1](https://user-images.githubusercontent.com/7538144/204928170-df67ed5e-7146-452b-a709-17b25458e991.png)
> > > Liên kết trỏ đến " [http://your-mooodle.com/local/bulkcustomcert/index.php?id={courseid}](http://your-mooodle.com/local/bulkcustomcert/index.php?id=%7Bcourseid%7D) " và bạn phải có khả năng "mod/customcert:viewallcertificates" để tải chúng xuống.
> > 
> > 
> > Xin chào@GonDragon tôi sử dụng moodle 4.1 sau đó đã cài đặt plugin của bạn, nhưng khi tôi kiểm tra thì không thể tìm thấy nút trên trang khóa học ngay cả khi sử dụng vai trò quản trị viên trang web. ![hình ảnh](https://private-user-images.githubusercontent.com/33312950/265868009-a1b3860c-2f69-4847-b852-510c60382f79.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0NTQ3NjQsIm5iZiI6MTcwNTQ1NDQ2NCwicGF0aCI6Ii8zMzMxMjk1MC8yNjU4NjgwMDktYTFiMzg2MGMtMmY2OS00ODQ3LWI4NTItNTEwYzYwMzgyZjc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE3VDAxMjEwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwODdkYjhiYjEwNWU2MGUyMDViY2RmM2VlNmZlZTVmMWNjYzQ3Y2M3NWZjZGNjZTBjYzMwOTFhNjcyYzhhNDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6iyv5uo1RoA7QcMPjupjZ6p1g8D8DfTolnj5aTURfB0)
> > Tôi mở liên kết bạn cung cấp trước đó (đã thay đổi trang moodle và id khóa học) không có gì xảy ra và nó chỉ quay lại trang khóa học. xin vui lòng tư vấn
> 
> I corrected a few places in @GonDragon code and the plugin worked very well. The problem is that there is confusion between role_assignments.id and users.id in code. I think it's probably due to the difference in Moodle version
> 
> If @GonDragon agrees I can send you the fix.

@ngonguyentuanhhon Could you share fix of @GonDragon code

**Comment by mdjnelson** [comment]
Since this seems to be a pretty popular request ill start coding a proper solution tomorrow. Stay tuned.

**Comment by mdjnelson** [comment]
Note - the feature isn't exactly what was suggested in the first comment. Downloading all the certificates site-wide would be a larger task. I only just realised when I re-read the first comment (this issue is quite old) that this isn't exactly what was asked for. In this task when you are viewing the list of users who have received a certificate you can also batch download them all into one zip.

**Comment by mdjnelson** [comment]
Done. Pushed to 4.2 and 4.1.

**Comment by ffdarkpenguin** [comment]
Hi @mdjnelson 

  is there any reason why it's not available in moodle 4.4? Or more important, will it be added to 4.4??? If yes, when?

Thanks,

Flávio

**Comment by mdjnelson** [comment]
I just am seeing your comment now @ffdarkpenguin as I am going to start work on https://github.com/mdjnelson/moodle-mod_customcert/issues/692 shortly. It should be available in 4.4 as it was introduced before that was even released.
